badger/badger.go

```diff
@@ -0,0 +1,346 @@
+	log   *logrus.Logger
+	mutex sync.Mutex
+}
+
+func (b *Store) Close(ctx context.Context) error {
+	return util.CallWithTimeout(ctx, b.DB.Close, func() {
+		b.log.Error("Closing of Badger Store timed out, Store may not shut down correctly.")
+	})
+}
+
+func (b *Store) Write(ctx context.Context, fn func(stoabs.WriteTx) error, opts ...stoabs.TxOption) error {
+	return b.doTX(ctx, func(tx *tx) error {
+		return fn(tx)
+	}, true, opts)
+}
+
+func (b *Store) Read(ctx context.Context, fn func(stoabs.ReadTx) error) error {
+	return b.doTX(ctx, func(tx *tx) error {
+		return fn(tx)
+	}, false, nil)
+}
+
+func (b *Store) WriteShelf(ctx context.Context, shelfName string, fn func(writer stoabs.Writer) error) error {
+	return b.doTX(ctx, func(tx *tx) error {
+		shelf, err := tx.GetShelfWriter(shelfName)
+		if err != nil {
+			return err
+		}
+		return fn(shelf)
+	}, true, nil)
+}
+
+func (b *Store) ReadShelf(ctx context.Context, shelfName string, fn func(reader stoabs.Reader) error) error {
+	return b.doTX(ctx, func(tx *tx) error {
+		shelf := tx.GetShelfReader(shelfName)
+		return fn(shelf)
+	}, false, nil)
+}
+
+func (b *Store) doTX(ctx context.Context, fn func(tx *tx) error, writable bool, opts []stoabs.TxOption) error {
+	if writable {
+		b.mutex.Lock()
+	}
+
+	// Start transaction, retrieve/create shelf to operate on
+	tx := &tx{
+		badgerTx: b.DB.NewTransaction(writable),
+		ctx:      ctx,
+		store:    b,
+	}
+	defer tx.rollback()
+
+	// Perform TX action(s)
+	appError := fn(tx)
+
+	// Writable TXs should be committed, non-writable TXs rolled back
+	if !writable {
+		tx.rollback()
+		return appError
+	}
+	// Observe result, commit/rollback
+	var err error
+	if appError == nil {
+		b.log.Trace("Committing Badger transaction")
+		// Check context cancellation, if not cancelled/expired; commit.
+		if ctx.Err() != nil {
+			err = ctx.Err()
+			tx.rollback()
+		} else if writable {
+			err = tx.commit()
+		}
+		b.mutex.Unlock()
+		if err != nil {
+			stoabs.OnRollbackOption{}.Invoke(opts)
+			return util.WrapError(stoabs.ErrCommitFailed, err)
+		}
+
+		stoabs.AfterCommitOption{}.Invoke(opts)
+	} else {
+		b.log.WithError(appError).Warn("Rolling back transaction application due to error")
+		tx.rollback()
+		stoabs.OnRollbackOption{}.Invoke(opts)
+		return appError
+	}
+
+	return nil
+}
+
+type tx struct {
+	ctx       context.Context
+	iterators []*badger.Iterator
+	mutex     sync.RWMutex
+	store     *Store
+	badgerTx  *badger.Txn
+}
+
+func (b *tx) Unwrap() interface{} {
+	return b.badgerTx
+}
+
+func (b *tx) GetShelfReader(shelfName string) stoabs.Reader {
+	return b.getBucket(shelfName)
+}
+
+func (b *tx) GetShelfWriter(shelfName string) (stoabs.Writer, error) {
+	return &badgerShelf{name: shelfName, tx: b}, nil
+}
+
+func (b *tx) getBucket(shelfName string) stoabs.Reader {
+	return &badgerShelf{name: shelfName, tx: b}
+}
+
+func (b *tx) Store() stoabs.KVStore {
+	return b.store
+}
+
+func (b *tx) newIterator() *badger.Iterator {
+	b.mutex.Lock()
```

Comment:
if locking here, why not in the other funcs of `tx` as well? Or is it a mutex for specific combinations of funcs (then it could use some comments)

Do we even allow concurrent usage of TXs? (sounds like a bad idea)

Reply:
comments added.
Badger panics when things go wrong. Better to prevent this.